provider_shopper/lib/screens/catalog.dart

```diff
@@ -0,0 +1,91 @@
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+import 'package:provider_shopper/models/cart.dart';
+import 'package:provider_shopper/models/catalog.dart';
+
+class MyCatalog extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: CustomScrollView(
+        slivers: [
+          _MyAppBar(),
+          SliverToBoxAdapter(child: SizedBox(height: 12)),
+          SliverList(
+            delegate: SliverChildBuilderDelegate(
+                (context, index) => _MyListItem(index)),
+          ),
+        ],
+      ),
+    );
+  }
+}
+
+class _AddButton extends StatelessWidget {
+  final Item item;
+
+  const _AddButton({Key key, @required this.item}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    var cart = Provider.of<CartModel>(context);
+
+    return FlatButton(
+      onPressed: cart.items.contains(item) ? null : () => cart.add(item),
+      splashColor: Theme.of(context).primaryColor,
+      child: cart.items.contains(item)
+          ? Icon(Icons.check, semanticLabel: 'ADDED')
+          : Text('ADD'),
+    );
+  }
+}
+
+class _MyAppBar extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return SliverAppBar(
+      title: Text('Catalog', style: Theme.of(context).textTheme.display4),
+      floating: true,
+      actions: [
+        IconButton(
+          icon: Icon(Icons.shopping_cart),
+          onPressed: () => Navigator.pushNamed(context, '/cart'),
+        ),
+      ],
+    );
+  }
+}
+
+class _MyListItem extends StatelessWidget {
+  final int index;
+
+  _MyListItem(this.index) : super(key: ValueKey(index));
```

Comment:
This isn't a method to create the key that I've seen elsewhere. Are there other widgets that do it? I'm used to the containing widget being responsible for creating and assigning the key.

Reply:
You're right. This was a remnant of some perf experiment. Got rid of it.